src/house.cpp

```diff
@@ -690,73 +739,155 @@ void Houses::payHouses(RentPeriod_t rentPeriod) const
 					case RENTPERIOD_DAILY:
-						period = "daily";
+						paidUntil += 24 * 60 * 60;
 						break;
-
 					case RENTPERIOD_WEEKLY:
-						period = "weekly";
+						paidUntil += 24 * 60 * 60 * 7;
 						break;
-
 					case RENTPERIOD_MONTHLY:
-						period = "monthly";
+						paidUntil += 24 * 60 * 60 * 30;
 						break;
-
 					case RENTPERIOD_YEARLY:
-						period = "annual";
+						paidUntil += 24 * 60 * 60 * 365;
 						break;
-
 					default:
 						break;
 				}
 
-				std::ostringstream ss;
-				ss << "Warning! \nThe " << period << " rent of " << house->getRent() << " gold for your house \"" << house->getName() << "\" is payable. Have it within " << daysLeft << " days or you will lose this house.";
-				letter->setText(ss.str());
-				g_game.internalAddItem(player.getInbox(), letter, INDEX_WHEREEVER, FLAG_NOLIMIT);
-				house->setPayRentWarnings(house->getPayRentWarnings() + 1);
+				house->setPaidUntil(paidUntil);
 			} else {
-				house->setOwner(0, true, &player);
+				if (house->getPayRentWarnings() < 7) {
+					int32_t daysLeft = 7 - house->getPayRentWarnings();
+
+					Item* letter = Item::CreateItem(ITEM_LETTER_STAMPED);
+					std::string period;
+
+					switch (rentPeriod) {
+						case RENTPERIOD_DAILY:
+							period = "daily";
+							break;
+
+						case RENTPERIOD_WEEKLY:
+							period = "weekly";
+							break;
+
+						case RENTPERIOD_MONTHLY:
+							period = "monthly";
+							break;
+
+						case RENTPERIOD_YEARLY:
+							period = "annual";
+							break;
+
+						default:
+							break;
+					}
+
+					std::ostringstream ss;
+					ss << "Warning! \nThe " << period << " rent of " << house->getRent() << " gold for your house \"" << house->getName() << "\" is payable. Have it within " << daysLeft << " days or you will lose this house.";
+					letter->setText(ss.str());
+					g_game.internalAddItem(player.getInbox(), letter, INDEX_WHEREEVER, FLAG_NOLIMIT);
+					house->setPayRentWarnings(house->getPayRentWarnings() + 1);
+				} else {
+					house->setOwner(0, true, &player);
+				}
 			}
-		}
+			IOLoginData::savePlayer(&player);
+
+		} else { // HOUSE_TYPE_GUILDHALL
+			Guild* guild = g_game.getGuild(ownerId);
+			if (!guild) {
+				guild = IOGuild::loadGuild(ownerId);
+				if (!guild) {
+					house->setOwner(0);
+					continue;
+				}
+			}
+
+			// If guild can afford paying rent
+			if (guild->getBankBalance() >= rent) {
+				guild->setBankBalance(guild->getBankBalance() - rent);
+
+				time_t paidUntil = currentTime;
+				switch (rentPeriod) {
+					case RENTPERIOD_DAILY:
+						paidUntil += 24 * 60 * 60;
+						break;
+					case RENTPERIOD_WEEKLY:
+						paidUntil += 24 * 60 * 60 * 7;
+						break;
+					case RENTPERIOD_MONTHLY:
+						paidUntil += 24 * 60 * 60 * 30;
+						break;
+					case RENTPERIOD_YEARLY:
+						paidUntil += 24 * 60 * 60 * 365;
+						break;
+					default:
+						break;
+				}
+
+				house->setPaidUntil(paidUntil);
+			} else { // guild cannot afford rent
+				Player player(nullptr);
```

Comment:
This code was copied from somewhere else, I might need to patch this in multiple locations. This is another sample of "c-style" code where variables are referenced through param `IOLoginData::loadPlayerById(&player, ownerId)`. 

![image](https://user-images.githubusercontent.com/1552144/85240048-fa5f3a80-b436-11ea-91a3-fc74f266a9a2.png)

Could this be performance related? Should I return tuples/pairs here? (bool and Player*)? Perhaps remaking this might be a little out of scope for this PR?

Reply:
This style seems to be the normal way of making sure the player you want is being loaded:
```cpp
Player* newOwner = g_game.getPlayerByGUID(guid);
if (!newOwner) {
	newOwner = new Player(nullptr);
	if (!IOLoginData::loadPlayerById(newOwner, guid)) {
		delete newOwner;
		return;
	}
}
```
As far as I understand, it first attempts to grab the player from g_game, which appears to hold data on players that have already been loaded into the game (online players perhaps)? And if that fails, use IOLoginData to retrieve and build the player from db. I think a separate PR (that is not related to guildhalls) could address something here, find a better way to do this. I think the above snippet is alot of code (just to find a player object).

Reply:
It would bot affect performance, but I would allocate and return an `std::unique_ptr<Player>` as it's more elegant and shows intent. Returning a pointer doesn't properly transfer ownership (as in should you delete this pointer? maybe it's managed somewhere else), a unique pointer is always owning.

However, I meant just use `Player player;` if possible.

Reply:
> However, I meant just to use `Player player;` if possible.

Tried this, but getting this error:
![image](https://user-images.githubusercontent.com/1552144/87251231-63503600-c46a-11ea-96b6-49094dbc5f04.png)

So it seems I have to modify the Player class to add support for this? I'm not good enough with pointers to figure this out.